### DISCUSSION TITLE
Add MediaElement Subtitle Support

### DISCUSSION THREAD

**Discussion by ne0rrmatrix** [CommunityToolkit/Maui#1867] [New Feature Discussions]
I would like to add support for subtitles for Android, Windows, Mac Catalyst and IOS. Supporting VTT and SRT file formats initially. Here is an example of what it could look like. I have been working on it for about a week and have some samples of how it might look. It still needs a lot of work. 

I am hoping everyone can vote and decide if we want this feature. I have a proposed API if it is approved. I was thinking we could just use: ```MediaElement.SubtitleUrl``` as the source for subtitles. This is open for any suggestions. I have yet to have a PR where the names of functions and or classes did not change a lot before it made it into a repo.
 
<p align="middle">
Android
</p>
<p align="middle">
<img src="https://github.com/CommunityToolkit/Maui/assets/4167863/0534626e-01ee-4d3a-b8b8-86d987e4bf29" width="350" align="top"  margin="10"/>
<img src="https://github.com/CommunityToolkit/Maui/assets/4167863/9b11966c-f15d-44ea-9989-af1ffd88e109" width="350" margin="10" />
</p>
<p align="middle">
Windows
</p>
<p align="middle">
<img src="https://github.com/CommunityToolkit/Maui/assets/4167863/a15355e9-939e-46ba-afa4-e3a172328bec" width="700" align="top"  margin="10"/>
<img src="https://github.com/CommunityToolkit/Maui/assets/4167863/8e6e2251-88bc-4925-abf4-faee5a565f0a" width="700" align="top"  margin="10"/>
</p>
 
### Proposed API ###
```csharp

/// <summary>
/// Backing store for the <see cref="SubtitleFont"/> property.
/// </summary>
public static readonly BindableProperty SubtitleFontProperty = BindableProperty.Create(nameof(SubtitleFont), typeof(string), typeof(MediaElement), string.Empty);

/// <summary>
/// Backing store for the <see cref="SubtitleFontSize"/> property.
/// </summary>
public static readonly BindableProperty SubtitleFontSizeProperty = BindableProperty.Create(nameof(SubtitleFontSize), typeof(double), typeof(MediaElement), 16.0);

/// <summary>
/// Backing store for the <see cref="SubtitleUrl"/> property.
/// </summary>
public static readonly BindableProperty SubtitleProperty = BindableProperty.Create(nameof(SubtitleUrl), typeof(string), typeof(MediaElement), string.Empty);

/// <summary>
/// Gets or sets the URL of the subtitle file to display.
/// </summary>
public string SubtitleUrl
{
    get => (string)GetValue(SubtitleProperty);
    set => SetValue(SubtitleProperty, value);
}

/// <summary>
/// Gets or sets the font to use for the subtitle text.
/// </summary>
public string SubtitleFont
{
    get => (string)GetValue(SubtitleFontProperty);
    set => SetValue(SubtitleFontProperty, value);
}

/// <summary>
/// Gets or sets the font size of the subtitle text.
/// </summary>
public double SubtitleFontSize
{
    get => (double)GetValue(SubtitleFontSizeProperty);
    set => SetValue(SubtitleFontSizeProperty, value);
}
```

**Comment by pictos** [comment]
How can we change the FontSize of the subtitle? On the last image it looks pretty tiny. And can you update your comment with the API spec?

**Reply by ne0rrmatrix** [reply]
That is a bug that I have already fixed. The size of the font did not change between the regular player and full screen causing an obvious issue with readability. I changed the font from 16 to 24 and it is a lot more readable. But the change was only applied when in full screen. 

### Current API ###

```csharp
/// <summary>
/// Backing store for the <see cref="SubtitleUrl"/> property.
/// </summary>
public static readonly BindableProperty SubtitleProperty = BindableProperty.Create(nameof(SubtitleUrl), typeof(string), typeof(MediaElement), string.Empty);

/// <summary>
/// Gets or sets the URL of the subtitle file to display.
/// </summary>
public string SubtitleUrl
{
	get => (string)GetValue(SubtitleProperty);
	set => SetValue(SubtitleProperty, value);
}
```
I am hoping for input on what we want to add to this?
Do we want to adjust font, font size? etc.

**Reply by pictos** [reply]
Yes, I would love to have the power to adjust the FontSize and make sure they follow the a11y font size that users set on their devices

**Reply by ne0rrmatrix** [reply]
Updated API

```csharp

/// <summary>
/// Backing store for the <see cref="SubtitleFont"/> property.
/// </summary>
public static readonly BindableProperty SubtitleFontProperty = BindableProperty.Create(nameof(SubtitleFont), typeof(string), typeof(MediaElement), string.Empty);

/// <summary>
/// Backing store for the <see cref="SubtitleFontSize"/> property.
/// </summary>
public static readonly BindableProperty SubtitleFontSizeProperty = BindableProperty.Create(nameof(SubtitleFontSize), typeof(double), typeof(MediaElement), 16.0);

/// <summary>
/// Backing store for the <see cref="SubtitleUrl"/> property.
/// </summary>
public static readonly BindableProperty SubtitleProperty = BindableProperty.Create(nameof(SubtitleUrl), typeof(string), typeof(MediaElement), string.Empty);

/// <summary>
/// Gets or sets the URL of the subtitle file to display.
/// </summary>
public string SubtitleUrl
{
    get => (string)GetValue(SubtitleProperty);
    set => SetValue(SubtitleProperty, value);
}

/// <summary>
/// Gets or sets the font to use for the subtitle text.
/// </summary>
public string SubtitleFont
{
    get => (string)GetValue(SubtitleFontProperty);
    set => SetValue(SubtitleFontProperty, value);
}

/// <summary>
/// Gets or sets the font size of the subtitle text.
/// </summary>
public double SubtitleFontSize
{
    get => (double)GetValue(SubtitleFontSizeProperty);
    set => SetValue(SubtitleFontSizeProperty, value);
}
```

**Comment by ne0rrmatrix** [comment]
IOS subtitle support is now working! 

https://github.com/CommunityToolkit/Maui/assets/4167863/ab3c3f09-ddb4-4ba4-a8db-f84efba15ab9

**Comment by ne0rrmatrix** [comment]
Windows, Android, and IOS all now working as expected. Android is dependant on merging #1873 
Layout is fully tested on Android. Further layout testing for IOS, Windows, and Mac required. No testing on Mac done yet. Will work on that this coming week.

**Reply by MuhammadTalhaahmed** [reply]
I'm really very excited and curious  for this 'Subtitle' feature, when will we get it in our MAUI APP ? Are we required to use the above API code OR the 'Subtitle' feature will be added in Media_Element Nuget package and will be easy to manage ?

**Reply by ne0rrmatrix** [reply]
This is a discussion where we talk about features we want in media element. In this case we are discussing subtitles and are trying to create an API for developers to use. Later if approved we will add a PR to repo and create the code that is needed based on the API specification. After it is reviewed by the team and accepted it can be merged. But and this is  big thing it needs to go through the whole approval process and then find a sponsor from the team to oversee the process from there.

**Comment by CliffAgius** [comment]
as there is a PR #1918 I am closing this discussion to clean up.

Let me know if it needs re-opening.
